core/templates/dev/head/library/Library.js

```diff
@@ -40,34 +40,72 @@ oppia.controller('Library', [
     // Keeps track of the index of the left-most visible card of each group.
     $scope.leftmostCardIndices = [];
 
-    $http.get('/libraryindexhandler').success(function(data) {
-      $scope.libraryGroups = data.activity_summary_dicts_by_category;
-
-      $rootScope.$broadcast(
-        'preferredLanguageCodesLoaded', data.preferred_language_codes);
-
-      $rootScope.loadingMessage = '';
-
-      // Pause is necessary to ensure all elements have loaded, same for
-      // initCarousels.
-      // TODO(sll): On small screens, the tiles do not have a defined width.
-      // The use of 214 here is a hack, and the underlying problem of the tiles
-      // not having a defined width on small screens needs to be fixed.
-      $timeout(function() {
-        tileDisplayWidth = $('exploration-summary-tile').width() || 214;
-      }, 20);
-
-      // Initialize the carousel(s) on the library index page.
-      $timeout(initCarousels, 390);
-
-      // The following initializes the tracker to have all
-      // elements flush left.
-      // Transforms the group names into translation ids
-      $scope.leftmostCardIndices = [];
-      for (i = 0; i < $scope.libraryGroups.length; i++) {
-        $scope.leftmostCardIndices.push(0);
-      }
-    });
+    $scope.inRankMode = ($window.location.pathname.indexOf(
+      '/explorations') === 0);
+    $scope.rankMethod = $window.location.pathname.substr(
+      $window.location.pathname.lastIndexOf('/') + 1);
+
+    if ($scope.inRankMode) {
+      $http.get('/librarygroupindexhandler/' + $scope.rankMethod).success(
+      function(data) {
+        $scope.libraryGroups = data.activity_summary_dicts_by_category;
+
+        $rootScope.$broadcast(
+          'preferredLanguageCodesLoaded', data.preferred_language_codes);
+
+        $rootScope.loadingMessage = '';
+
+        // Pause is necessary to ensure all elements have loaded, same for
+        // initCarousels.
+        // TODO(sll): On small screens, the tiles do not have a defined width.
+        // The use of 214 here is a hack, and the underlying problem of the
+        // tiles not having a defined width on small screens needs to be fixed.
+        $timeout(function() {
+          tileDisplayWidth = $('exploration-summary-tile').width() || 214;
+        }, 20);
+
+        // Initialize the carousel(s) on the library index page.
+        $timeout(initCarousels, 390);
+
+        // The following initializes the tracker to have all
+        // elements flush left.
+        // Transforms the group names into translation ids
+        $scope.leftmostCardIndices = [];
+        for (i = 0; i < $scope.libraryGroups.length; i++) {
+          $scope.leftmostCardIndices.push(0);
+        }
+      });
+    } else {
+      $http.get('/libraryindexhandler').success(
```

Comment:
Oh, _interesting_ -- I just realized that this handler gets called way more than it needs to be. It even gets called during the search results, when the library groups are not actually showing. Perhaps we should put "else if (!$scope.searchMode)" here.